chainer/training/extensions/extractor.py

```diff
@@ -0,0 +1,199 @@
+import copy
+
+import six
+
+from chainer import cuda
+from chainer.dataset import convert
+from chainer.dataset import iterator as iterator_module
+from chainer import link
+from chainer.training import extension
+from chainer import variable
+
+
+def concat_variables(variables, device=None):
+    """Concatenates a list of variables into a single variable.
+
+    The forward operation of a model yields a variable, tuple of variables
+    or a dict of variables for each batch. This function concatenates a list
+    of such variables to a single variable, tuple of varialbes or dict of
+    variables, respectively.
+
+    Args:
+        variables (list): A list of variables, list of tuples of variables
+            or a list of dicts of variables.
+        device (int): Device ID to which each array is sent. Negative value
+            indicates the host memory (CPU). If it is omitted, all arrays are
+            left in the original device.
+
+    Returns:
+        Variable, a tuple of variables, or a dictionary of variables. The type
+        depends on the type of each example in the variables list.
+
+    """
+    if len(variables) == 0:
+        raise ValueError('No variables given')
+
+    if device is None:
+        def to_device(x):
+            return x
+    elif device < 0:
+        to_device = cuda.to_cpu
+    else:
+        to_device = lambda x: cuda.to_gpu(x, device)
+
+    first_elem = variables[0]
+
+    if isinstance(first_elem, tuple):
+        result = []
+        for i in six.moves.range(len(first_elem)):
+            result.append(to_device(_concat_variables(
+                [example[i] for example in variables])))
+
+        return tuple(result)
+
+    elif isinstance(first_elem, dict):
+        result = {}
+        for key in first_elem:
+            result[key] = to_device(_concat_variables(
+                [example[key] for example in variables]))
+
+        return result
+
+    else:
+        return to_device(_concat_variables(variables))
+
+
+def _concat_variables(variables):
+    xp = cuda.get_array_module(variables[0].data)
+    with cuda.get_device(variables[0].data):
+        return variable.Variable(
+            xp.concatenate([var.data for var in variables])
+        )
+
+
+class Extractor(extension.Extension):
+
+    """Trainer extension to extract features using a trained model.
+
+    This extension extracts features using a trained model by a given
+    extraction function.
+
+    Extractor has a structure to customize similar to that of
+    :class:`~chainer.training.extensions.Evaluator`.
+
+    Args:
+        iterator: Dataset iterator for extracting features. It can also be
+            a dictionary of iterators. If this is just an iterator, the
+            iterator is registered by the name ``'main'``.
+        target: Link object or a dictionary of links to process. If this is
```

Comment:
In case of `target` being a dictionary, only target['main'] is ever used (rest are ignored). Does it make sense to support a dictionary of targets as possible input for the purposes of feature extraction? In that case features might have to be a dictionary indexable by target name. In case no, only accept `target` to be a link.Link